src/doc/rustc/src/exploit-mitigations.md

```diff
@@ -0,0 +1,588 @@
+   <td>0.12.0 (2014-10-09)
+   </td>
+  </tr>
+  <tr>
+   <td>Integer overflow checks
+   </td>
+   <td>Supported but enabled by default when debug assertions are enabled only.
+   </td>
+   <td>1.1.0 (2015-06-25)
+   </td>
+  </tr>
+  <tr>
+   <td>Non-executable memory regions
+   </td>
+   <td>Yes
+   </td>
+   <td>1.8.0 (2016-04-14)
+   </td>
+  </tr>
+  <tr>
+   <td>Stack clashing protection
+   </td>
+   <td>Yes
+   </td>
+   <td>1.20.0 (2017-08-31)
+   </td>
+  </tr>
+  <tr>
+   <td>Read-only relocations and immediate binding
+   </td>
+   <td>Yes
+   </td>
+   <td>1.21.0 (2017-10-12)
+   </td>
+  </tr>
+  <tr>
+   <td>Heap corruption protection
+   </td>
+   <td>Yes
+   </td>
+   <td>1.32.0 (2019-01-17) via OS default or specified allocator
+   </td>
+  </tr>
+  <tr>
+   <td>Stack smashing protection
+   </td>
+   <td>No
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>Forward-edge control flow protection
+   </td>
+   <td>No
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>Backward-edge control flow protection (e.g., shadow and safe stack)
+   </td>
+   <td>No
+   </td>
+   <td>
+   </td>
+  </tr>
+</table>
+
+<small id="fn:1">1\. See
+<https://github.com/rust-lang/rust/tree/master/compiler/rustc_target/src/spec>
+for a list of targets and their default options. <a href="#fnref:1"
+class="reversefootnote" role="doc-backlink">↩</a></small>
+
+
+### Position-independent executable
+
+Position-independent executable increases the difficulty of the use of code
+reuse exploitation techniques, such as return-oriented programming (ROP) and
+variants, by generating position-independent code for the executable, and
+instructing the dynamic linker to load it similarly to a shared object at a
+random load address, thus also benefiting from address-space layout
+randomization (ASLR). This is also referred to as “full ASLR”.
+
+The Rust compiler supports position-independent executable and enables it by
+default since version 0.12.0 (2014-10-09)[10]–[13].
+
+```
+$ readelf -h target/release/hello-rust | grep Type:
+  Type:                              DYN (Shared object file)
+```
+Fig. 1. Checking if an executable is a position-independent executable.
+
+An executable with an object type of `ET_DYN` (i.e., shared object) and not
+`ET_EXEC` (i.e., executable) is a position-independent executable (see Fig.
+1).
+
+
+### Integer overflow checks
+
+Integer overflow checks protects programs from undefined and unintended
+behavior (which may cause vulnerabilities) by checking for results of signed
+and unsigned integer computations that cannot be represented in their type,
+resulting in an overflow or wraparound.
+
+The Rust compiler supports integer overflow checks, and enables it when
+debug assertions are enabled since version 1.0.0 (2015-05-15)[14]–[17], but
+support for it was not completed until version 1.1.0 (2015-06-25)[16]. An
+option to control integer overflow checks was later stabilized in version
+1.17.0 (2017-04-27)[18]–[20].
```

Comment:
Thanks for the history, but I don't think it is needed to including this.
We want to follow what current nightly do.

Reply:
Yes, in general, all docs are written "in the present", and we don't include specific version numbers.

Reply:
I can remove these. However, this entire document is written as a historical account of these exploit mitigations' major milestones for Rust, and removing these will make this part inconsistent with the rest of the document. Let me know if you still prefer I remove these.

Reply:
Yes, we don't write docs in that style. That doesn't mean it's not useful to have, it's just not the way that the documentation is organized.